Initial API proposal

jsternberg opened the pull request:
# Initial API Proposal

This initial proposal needs to just be able to do two simple things. It needs to be able to query the database and write to the database.
## Queries

The basic query interface has been modified to just returning the body from the request and building additional parsing functionality on top of that. I thought about creating some kind of response struct or interface, but there were a few different use cases that I had to consider.
1. Multiple content types so there are multiple ways for a server to return data depending on what format the client asks for.
2. Some content types we will just want the client to print verbatim. JSON and CSV can be printed directly and we will continue doing that with new content types as time goes on. The column format is the only one that will not be done on the server. A client who wants to do that will choose another format and decode the format it asks for. An error will be returned if the server returns a non-200 OK status message or it will return an empty reader if 204 OK was returned instead.
3. We will have convenience methods for parsing the resulting stream.

``` go
c := &influxdb.Client{}
query, meta, err := c.RawQuery("SHOW DATABASES", &influxdb.QueryOptions{Format: "json"})
if err != nil {
    return nil
}

reader, err := influxdb.NewReader(query, meta)
// read from the query
```

This typically gets encapsulated into a `Query()` method that handles this. The user gets access to either make a `RawQuery()` or to make a `Query()` when they want to parse the results. We can change which format we request with `Query()` to the most efficient one we can (such as switching it to msgpack in the future) without any kind of negative.
## Writer

We will want to support multiple methods of writing data. The HTTP client is the most powerful and it needs to support both TCP and UNIX sockets. While UDP is there, it will only accept a hostname and does not need to use a `influxdb.Client` struct since it has no connection to that piece of code. All batch writing interfaces are unified with an `influxdb.Writer` interface along with just having a convenience method on the client itself. The `influxdb.Writer` assumes it will only be writing to one database or retention policy. If they need to write to multiple databases or retention policies, they will need to make multiple writers.

jsternberg commented:
@influxdata/core my initial proposal for the new API. We would include more convenience functions later (such as convenience methods to parse the JSON response), but I wanted to get the initial parts done first just to start a conversation.

corylanou commented on client.go:
```diff
@@ -0,0 +1,165 @@
+	case PrecisionHour:
+		return "h"
+	}
+	return ""
```
should be a default with a panic?  Unreachable code right?

jsternberg replied:
It's technically reachable if you do something like this: `influxdb.Precision(-1).String()` and I thought returning an empty string would be better than panicking.

corylanou replied:
This seems unnecessarily complex to me.  I would just use constants with the actual values.  And then validate when you issue queries/writes and return errors accordingly.  The goal is to prevent programmers from doing dumb things.

corylanou commented on client.go:
```diff
@@ -0,0 +1,165 @@
+
+const (
+	// ConsistencyAll requires all data nodes to acknowledge a write.
+	ConsistencyAll = iota
```
any reason to do these as itoa and not just the string type and be done?  Same with precision?

jsternberg replied:
Probably not, but it is useful for default values. The way things are currently set up, the top value for `iota` is the default value for the client to use when one isn't set. I should check to see if the server sets a default though because maybe we just want to send an empty string then.

nathanielc commented:
@jsternberg Since this is yet another re-write of the client what are we doing differently from the previous clients? I was able to pick up on a few things from your description above about different content types etc, but what else? In other words what problems are we trying to solve?

For example, in Kapacitor I had to write my own client so that I could add special authentication headers and such to the HTTP requests. If the v2 client had exposed the request object then I could have used it instead.

jsternberg commented:
The main difference I had in mind was to expose some parts in the API that weren't previously exposed but to also try to make it fit with how I viewed the API should be used. A primary purpose I had was making it easier to create and write points and to get the raw query result from the server rather than forcing the user to wait for it to parse the entire response. The latter was especially needed since consuming any large amount of input was impossible with the old structure.

Can you give some more detail about what you did? If it's something that should be built into the library, we should include that authentication in the library itself. If it's something that was special to Kapacitor, maybe we should try exposing the HTTP request objects so the user could modify them if they wanted to.

nathanielc commented:
@jsternberg You can find the implementation here: https://github.com/influxdata/kapacitor/tree/master/influxdb

It is basically a copy of the existing v2 client but with changes that allowed me to set auth credentials. 

> expose some parts in the API that weren't previously exposed

What parts of the API aren't exposed vai the v2 client?

> try to make it fit with how I viewed the API should be used

Can you expound on this?

jsternberg commented:
> What parts of the API aren't exposed vai the v2 client?

Chunking wasn't exposed at all and it's not possible to retrieve back alternate content types like CSV or the currently pending MsgPack format. The client was made before either of those existed and chunking wasn't fixed and actively implemented until after the v2 client existed. Since the response is read by the client into a single struct that it returns, it's impossible to implement something like this:

```
client := influxdb.Client{}

// This next command is uninterruptible.
resp, err := client.Query("SELECT * FROM cpu")
if err != nil {
    log.Println(err)
    return
}
```

The new API allows you to interrupt this, although I think it requires a fix to the server that I can't remember if it's merged or not.

```
client := influxdb.Client{}
r, meta, err := client.RawQuery("SELECT * FROM cpu")
if err != nil {
    log.Println(err)
    return
}

// now you can read the response and you can close the reader if you want to interrupt early.
```

jsternberg commented:
> try to make it fit with how I viewed the API should be used

I forgot to include the response to this in my previous comment.

So it's mostly I wanted to try and make the API more simple and account for more common use cases. The query section is the part that I really felt had to be changed, but the write section also has some notable difference. Instead of having just one write method and having it take a batch of points, I thought a writer interface would be more appropriate and that way we could build things like the proposed BufferedWriter on top of it. I also wanted to include support for UDP and Unix sockets which the previous client doesn't really have a method of exposing. I'm not completely sure on the writer interface though as we may want to make it possible to write raw line protocol which wouldn't currently be possible.

jsternberg commented:
@nathanielc thinking about some of the things you have said, do you think it would be reasonable to open up certain portions of the API more? Something like:

```
func NewQuery(q string, opt *QueryOptions) *http.Request {
    // construct the request and return it
}
```

This way, a person could make whatever modifications they wanted to the request before it gets sent off to the `client.Do` call. That's technically already exposed so it may be useful for certain use cases that we may not anticipate. Similarly, we should allow the same with the writer too.

desa commented:
Just went through our repos looking for places that write/query the DB:

# Writing
* [v1 stress point generator](https://github.com/influxdata/influxdb/blob/master/stress/basic.go#L129-L138)
* [v2 stress point generator](https://github.com/influxdata/influxdb/blob/master/stress/v2/statement/insert.go#L105-L153)
* [kapacitor](https://github.com/influxdata/kapacitor/blob/master/influxdb/client.go#L241-L251)
* [telegraf](https://github.com/influxdata/telegraf/blob/master/plugins/outputs/influxdb/influxdb.go#L193-L195). (Note: the [`telegraf.Metric`](https://github.com/influxdata/telegraf/blob/master/metric.go#L26) type is used to generate a `client.Point`, but could just as easily product line protocol.)
* [mountainflux](https://github.com/mark-rushakoff/mountainflux) (The point generation is done in [river](https://github.com/mark-rushakoff/mountainflux/blob/master/river/river.go) and the writing is done in [avalanche](https://github.com/mark-rushakoff/mountainflux/tree/master/avalanche).)
* [laborator](https://github.com/influxdata/laboratory)
  - [Generator interface](https://github.com/influxdata/laboratory/blob/master/experiment/generator.go)
  - [counter implementation](https://github.com/influxdata/laboratory/blob/master/generators/counter.go) of the Generator interface
   - [another example imeplementation](https://github.com/influxdata/laboratory/blob/master/experiment/distinctduration/generator.go) of the Generator interface.
* [influxdb.lab (python)](https://github.com/influxdata/influxdb.lab/blob/master/influxdb/lab/target.py#L109-L209)

# Querying
* [Chronograf Reverse Proxy](https://github.com/influxdata/chronograf/blob/master/server/reverse_proxy_pool.go) with [templating](https://github.com/influxdata/chronograf/tree/master/statement)
* [laboratory](https://github.com/influxdata/laboratory/tree/master/resultparser) and its associated [querier](https://github.com/influxdata/laboratory/tree/master/queriers) and an [example of its use](https://github.com/influxdata/laboratory/blob/master/experiment/timetonpoints/time_to_n_points.go#L94-L106) and [another example](https://github.com/influxdata/laboratory/blob/548b9a00e93278e1eb269841720bcb26eb9c379b/resultparser/result_parser_test.go)
* [influxdb.lab Querier](https://github.com/influxdata/influxdb.lab/blob/master/influxdb/lab/target.py#L261-L285)

jsternberg commented:
I just added some unit tests that pretty much all fail and I'm going to continue working on them. I also modified some of the method names and added some additional functionality. Notable things:

Query has been extracted into multiple methods and the function signature changed. There is now a `Select()` method whose primary purpose is for returning a Reader. This was the old `Query` method. I also added an `Execute()` method. This method differs because it discards all output except it uses the Reader to look for any errors. The Reader has also been modified to allow passing a `nil` Result. If the Result is nil, it will only attempt to decode the error and will discard any other results. The rationale for this is that creating a database and selecting from a database are usually two separate things people do even if we use the same endpoint for both and there is usually different desire when performing both. When attempting to create a database, you are going to discard the output anyway and only want to read the output to see if there was an error returned. Select will also issue a `GET` and Execute will issue a `POST` by default so that Select is read only.

The `NewQuery` function is now exposed. It constructs a new `http.Request` struct that can be then sent to `client.Do`. This way you can customize the HTTP request. The raw query function has been removed and you should instead just do something like this:

```go
client := influxdb.Client{}
query, _ := client.NewQuery("SELECT * FROM cpu", nil)
resp, err := client.Do(query)
...
```

You can then read from the response body directly. NewQuery defaults to using `POST`, but maybe we should expose the read only stuff so that it can setup the `GET` or `POST` correctly too.

The parameter has been changed to `interface{}` instead of `string`. I remembered we have the ability to upload files now for the query. This was primarily so that people could upload files with curl which will send the file as a `multipart/form-data`, but this kind of input is better when people want to upload something like an initial database creation that spans multiple commands. If the parameter is a string, it will use the old behavior I documented. If the input is an io.Reader, it will instead upload that as `multipart/form-data` and use `POST`.

I modified `Select()`, but now that I think about it, `GET` and `multipart/form-data` are incompatible. Should I restrict `Select()` to only accept a string or should I have `Select()` automatically read the `io.Reader` and encode the query into the parameter string? I'm thinking the former is correct.

mark-rushakoff commented:
@desa and I have implemented writers and queriers in several load-generation cases across multiple programming languages. We discussed client API design for a bit yesterday, and here's some of our conclusions based on our real world experience writing into and querying from InfluxDB.

## Writers

Writing should expose a low-level API where the consumer specifies

* Line protocol as a byte slice
* Time precision
* Write consistency 
* Target database and retention policy
* **Retry strategy** – 40X status codes indicate a client error like specifying a database that doesn't exist or using an incompatible field type. But 500s *usually* mean something is okay to retry, e.g. a partial write. Logic around retries is complicated, and API consumers do not want to be concerned with the small details.

Once you've established a solid API that works directly with line protocol, the API consumer is free to use other abstractions that we provide, or their own abstractions, to generate and write points. 

@desa has some partial work somewhere that allows you to use struct tags and uses reflection to automatically generate line protocol from a struct you can control.  It would be easy to plug in to anything that accepted line protocol.

The `Point` model as exists right now, would need to re-serialize the tags for each point inserted; this is wasteful for clients who know they're only ever writing a small number of series with a large number of tags, for instance. But if someone had code depending on the `Point` struct already, they should be able to easily use  any client that accepted line protocol.

## Queriers

While our query results in JSON are network-efficient, the separation between columns and values is a constant annoyance when consuming query results. You either have to assume the order of the columns (which may be okay for queries you've constructed but is bad for meta-queries like `SHOW STATS`) or you have to map the columns to each value entry. And then you have to typecast - hopefully you know the field types up front!

Although there's a bit of mental overhead in this model, I've found that specifying "I am looking for a column of this type and this name" is a very pleasant API. I pulled out the code from the not-yet-public laboratory repo into [this gist](https://gist.github.com/mark-rushakoff/2e1c160a15d0c7500548d9aae1773c6d) for anyone who wants to see an example of it. The downside with this implementation is that it only works with a single result with a single series, not multiple. And it probably doesn't handle error responses properly.

Providing a callback function to a result parser is another option.  Perhaps the callback signature could be something like `func (resultIdx, seriesIdx, valueIdx int, tags map[string]string, values map[string]interface{})`. I don't like the client having to control all the typecasting on each value row, and I don't like how it's inconvenient to handle query errors either.

## Are writers and queriers the same?

I don't think they are. I think _most_ applications are either write-only (exception for query to create database) or query-only. Applications that write *and* query are candidates to implement in Kapacitor.

I would like to see an API that properly distinguishes writes and queries. Response content, retry strategies, and error handling cases are significantly different.

nathanielc commented:
> I would like to see an API that properly distinguishes writes and queries. Response content, retry strategies, and error handling cases are significantly different.

I agree with this. As Kapacitor does both, but they are completely different sections of the code base.

jsternberg commented:
@mark-rushakoff I think my latest iteration may have hit a lot of your points, but not all of them.

Write protocol is now abstracted away from the `client.Write()` method. Now you write the points to a buffer and that buffer gets attached to the request so you can choose to use more low-level write methods for sending a write. I think we might need to separate the retry logic for the write though into an alternate method though so there will be three parts.

1. Write the points to a buffer.
2. Create HTTP request.
3. Send HTTP request with retry logic.

The `client.Write()` function was supposed to do all three, but only 1 and 2 are exposed right now if you want to implement it yourself. It might be worth also exposing 3.

I also moved the line protocol generation into the `Point.Write(io.Writer)` method instead. Tags are kept in an array (not a map) to facilitate quick writing, but you're right that it might be worth creating a way to pre-compile a byte slice for these that can then be used. Maybe I should turn line protocol writing into an interface so that `Point` can write line protocol in the common case, but we can then have other convenience writers that pre-compile the beginning. Although if we expose all parts of the writer, it's not very difficult for these special writers to write to a buffered stream themselves.

For queries, I added `ForEach` to be able to use a callback function. I do think we should add some convenience functions to be able to index into results. In the test cases I've written, they're a huge PITA to be able to actually index into anything you want.

Can you give some feedback on the API as I've currently got it and I'll try to address some of your concerns too.

mark-rushakoff commented on client.go:
```diff
@@ -0,0 +1,158 @@
+
+// NewClient creates a new client pointed to the parsed hostname.
+func NewClient(url string) (*Client, error) {
+	proto, addr, path, err := ParseUrl(url)
```
May need special logic to handle basic auth. It's extremely convenient to initialize a client as `http://user:password@influx.example.com:8086`. And this may require a `url.UserInfo`, or separate `Username` and `Password` strings.

jsternberg replied:
It might be worth just not having `ParseUrl` since it's not particularly useful outside of this. `url.ParseUrl` is likely good enough.

mark-rushakoff commented on client.go:
```diff
@@ -0,0 +1,158 @@
+// application/x-www-form-urlencoded. If the query is an io.Reader, the query is sent
+// as a file using multipart/form-data. The first is more useful for a single ad-hoc
+// query, but the second can be better for running large multi-command queries.
+func (c *Client) NewQuery(method string, q interface{}, opt *QueryOptions) (*http.Request, error) {
```
I like the idea of a request abstraction. I'm slightly unsure of returning a straight `http.Request` because I don't know if it's okay to commit to HTTP (vs. UDP or other transports), and I don't know if a `http.Request` struct is sufficient or if we need additional information to execute the request. 

It seems like it should be a struct belonging to this package that you pass to another method on `Client`. If it's meant to be passed to `Select` and `Execute`, then I think it would be more developer-friendly to use our own struct and have our own, non-empty interface for the `q` arguments to those methods.

jsternberg replied:
I'm not sure if we need an abstraction or not. I would be open to one in the case that we do something like TCP, but I also think at that point we'll probably create a new API since the API would have diverged enough. This is a valid point though and I'll see what I can do. I'm just afraid of not offering enough knobs when people really need them.

I don't think we're going to ever switch to UDP though. UDP doesn't offer the ability to get a response back and isn't really appropriate for this type of API. I do think it's possible we might create a binary TCP response in the future, but that's also unlikely in the short term (and maybe in the long term).

mark-rushakoff commented on client.go:
```diff
@@ -0,0 +1,158 @@
+
+// Execute executes a query, discards the results, and returns any error that may have happened.
+// If the error happened as a result of a statement failing for some reason, the error is wrapped
+// in a ResultError.
```
Comment should mention you would use Execute for "meta" queries like create database.

mark-rushakoff commented on reader.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+// SameSeries returns if this is the same series as the other series.
+func (s *Series) SameSeries(o *Series) bool {
```
When would a client compare series?

jsternberg replied:
This is likely the wrong abstraction and we should work on that, but chunking means that a series can be split between multiple results and we need some easy way for a person to identify "this is the same series as the one that came before". There's also a bit of a problem with our current API output since it's not possible to know from the first request if it is a partial series or not. I think we can improve the API and integrate that with the client at some point, but I haven't run that change across anybody yet and it definitely won't be in 1.0.

mark-rushakoff commented:
I went through the code a bit and left a couple comments, but it's difficult to add meaningful commentary about a library when only relying on unit tests as references of use.

I think the most useful thing to do at this point would be to actually use the client library in a couple small example apps. For writes, perhaps a simple echo server could track source IP/port, duration of connection being open, length of message sent, etc. For queries, you could look at that same data and generate a couple reports of average latency over past 1/5/15 minutes, most frequent source IPs, etc.

jsternberg commented:
Not going with this proposal. I just merged #3 as the initial query API.
